src/MaxMind/Db/Reader/Decoder.php

```diff
@@ -197,105 +221,107 @@ private function decodeMap($size, $offset)
+        switch ($pointerSize) {
+            case 1:
+                $packed = (pack('C', $ctrlByte & 0x7)) . $buffer;
+                list(, $pointer) = unpack('n', $packed);
+                $pointer += $this->pointerBase;
+                break;
+            case 2:
+                $packed = "\x00" . (pack('C', $ctrlByte & 0x7)) . $buffer;
+                list(, $pointer) = unpack('N', $packed);
+                $pointer += $this->pointerBase + 2048;
+                break;
+            case 3:
+                $packed = (pack('C', $ctrlByte & 0x7)) . $buffer;
+
+                // It is safe to use 'N' here, even on 32 bit machines as the
+                // first bit is 0.
+                list(, $pointer) = unpack('N', $packed);
+                $pointer += $this->pointerBase + 526336;
+                break;
+            case 4:
+                // We cannot use unpack here as we might overflow on 32 bit
+                // machines
+                $pointer = $this->decodeUint($buffer, $pointerSize, $this->pointerBase);
+        }
 
         return [$pointer, $offset];
     }
 
-    private function decodeUint($bytes)
+    private function decodeUint($bytes, $byteLength, $base)
     {
-        list(, $int) = unpack('N', $this->zeroPadLeft($bytes, 4));
-
-        return $int;
-    }
-
-    private function decodeBigUint($bytes, $byteLength)
-    {
-        $maxUintBytes = log(PHP_INT_MAX, 2) / 8;
-
         if ($byteLength === 0) {
-            return 0;
+            return $base;
         }
 
-        $numberOfLongs = ceil($byteLength / 4);
-        $paddedLength = $numberOfLongs * 4;
-        $paddedBytes = $this->zeroPadLeft($bytes, $paddedLength);
-        $unpacked = array_merge(unpack("N$numberOfLongs", $paddedBytes));
-
         $integer = 0;
 
-        // 2^32
-        $twoTo32 = '4294967296';
+        for ($i = 0; $i < $byteLength; ++$i) {
+            $part = \ord($bytes[$i]);
 
-        foreach ($unpacked as $part) {
             // We only use gmp or bcmath if the final value is too big
-            if ($byteLength <= $maxUintBytes) {
-                $integer = ($integer << 32) + $part;
+            if ($byteLength <= _MM_MAX_INT_BYTES) {
+                $integer = ($integer << 8) + $part;
             } elseif (\extension_loaded('gmp')) {
-                $integer = gmp_strval(gmp_add(gmp_mul($integer, $twoTo32), $part));
+                $integer = gmp_add(gmp_mul($integer, 256), $part);
             } elseif (\extension_loaded('bcmath')) {
-                $integer = bcadd(bcmul($integer, $twoTo32), $part);
+                $integer = bcadd(bcmul($integer, 256), $part);
             } else {
                 throw new \RuntimeException(
                     'The gmp or bcmath extension must be installed to read this database.'
                 );
             }
         }
 
-        return $integer;
-    }
+        if ($base > 0) {
+            $byteLength = $byteLength + log($base, 2) / 8;
```

Comment:
Should we use `ceil()` or `+ 1` on the `log()` calculation? e.g. if `$base` was 1024. I suppose I'm just worried about edge cases with the floating point here. Possibly we'd want that on the other `log()` use above as well although maybe `floor()` would be more appropriate there?

Reply:
Hmm, I am having a hard time imagining where this would be an issue. Is your concern limited to just a floating point math situation where `$byteLength + log($base, 2) / 8` should be actually bigger than `log(PHP_INT_MAX, 2) / 8` but is smaller or equal? I guess the "or equal" part is a bit concerning, although i am not convinced it can happen with these numbers. However, reviewing the code, it looks like we just need this in two places. Maybe I'll get rid of the base stuff so that we can precompute the numbes.

Reply:
Assuming this does happen, wouldn't `ceil` have the same issue if log returned an integer?

Reply:
My thinking was mainly whether the floating point part could lead to the wrong branch. I agree it is hard to see it being a problem. I was also thinking conceptually about fractional bytes being potentially not obvious as the first use wouldn't be making use of that. Leaving it or decrementing a little seems fine!